Run tests via setup.py, and mark and skip by default slow tests. Enable these tests in Travis-CI

kinow opened the pull request:
This PR was supposedly to be well tested today, but alas I spent a considerable time battling Travis-CI. First with weird issues that required to remind it to always update `pip` and `setuptools`, as it was unhappy with the version of the package.

Then it started building, but it took over 30 minutes in two of the 4 build jobs. Not sure if that's just Travis-CI acting funny, or something introduced in these changes. But I finished the changes, and left it running and decided to open the pull request anyway to finish my code review in the pull request today, and so that you could take a look and see what you think about it.

This way you should be able to run `python setup.py test` to run the tests, except the two that take a while (and make my notebook's fan to start up). These tests should finish under 10 seconds.

```bash
$ python3 setup.py test 
running pytest
running egg_info
writing isodatetime.egg-info/PKG-INFO
writing dependency_links to isodatetime.egg-info/dependency_links.txt
writing entry points to isodatetime.egg-info/entry_points.txt
writing top-level names to isodatetime.egg-info/top_level.txt
reading manifest file 'isodatetime.egg-info/SOURCES.txt'
writing manifest file 'isodatetime.egg-info/SOURCES.txt'
running build_ext
=============================================================================================== test session starts ===============================================================================================
platform linux -- Python 3.7.0, pytest-3.8.0, py-1.6.0, pluggy-0.7.1 -- /home/kinow/Development/python/anaconda3/bin/python3
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/home/kinow/Development/python/workspace/isodatetime/.hypothesis/examples')
rootdir: /home/kinow/Development/python/workspace/isodatetime, inifile: pytest.ini
plugins: remotedata-0.3.0, openfiles-0.3.0, env-0.6.2, doctestplus-0.1.3, cov-2.6.0, arraydiff-0.2, hypothesis-3.83.1
collected 48 items                                                                                                                                                                                                

isodatetime/tests/test_00.py::TestSuite::test_days_in_year_range SKIPPED                                                                                                                                    [  2%]
isodatetime/tests/test_00.py::TestSuite::test_duration_floordiv PASSED                                                                                                                                      [  4%]
isodatetime/tests/test_00.py::TestSuite::test_duration_in_weeks_floordiv PASSED                                                                                                                             [  6%]
isodatetime/tests/test_00.py::TestSuite::test_duration_subtract PASSED                                                                                                                                      [  8%]
isodatetime/tests/test_00.py::TestSuite::test_duration_to_weeks PASSED                                                                                                                                      [ 10%]
isodatetime/tests/test_00.py::TestSuite::test_get_local_time_zone_format PASSED                                                                                                                             [ 12%]
isodatetime/tests/test_00.py::TestSuite::test_get_local_time_zone_no_dst PASSED                                                                                                                             [ 14%]
isodatetime/tests/test_00.py::TestSuite::test_get_local_time_zone_with_dst PASSED                                                                                                                           [ 16%]
isodatetime/tests/test_00.py::TestSuite::test_largest_truncated_property_name PASSED                                                                                                                        [ 18%]
isodatetime/tests/test_00.py::TestSuite::test_smallest_missing_property_name PASSED                                                                                                                         [ 20%]
isodatetime/tests/test_00.py::TestSuite::test_timeduration PASSED                                                                                                                                           [ 22%]
isodatetime/tests/test_00.py::TestSuite::test_timeduration_add_week PASSED                                                                                                                                  [ 25%]
isodatetime/tests/test_00.py::TestSuite::test_timeduration_dumper PASSED                                                                                                                                    [ 27%]
isodatetime/tests/test_00.py::TestSuite::test_timeduration_parser PASSED                                                                                                                                    [ 29%]
isodatetime/tests/test_00.py::TestSuite::test_timepoint_add_duration PASSED                                                                                                                                 [ 31%]
isodatetime/tests/test_00.py::TestSuite::test_timepoint_add_duration_without_minute PASSED                                                                                                                  [ 33%]
isodatetime/tests/test_00.py::TestSuite::test_timepoint_dumper PASSED                                                                                                                                       [ 35%]
isodatetime/tests/test_00.py::TestSuite::test_timepoint_dumper_after_copy PASSED                                                                                                                            [ 37%]
isodatetime/tests/test_00.py::TestSuite::test_timepoint_dumper_bounds_error_message PASSED                                                                                                                  [ 39%]
isodatetime/tests/test_00.py::TestSuite::test_timepoint_dumper_get_time_zone PASSED                                                                                                                         [ 41%]
isodatetime/tests/test_00.py::TestSuite::test_timepoint_parser PASSED                                                                                                                                       [ 43%]
isodatetime/tests/test_00.py::TestSuite::test_timepoint_plus_float_time_duration_day_of_month_type PASSED                                                                                                   [ 45%]
isodatetime/tests/test_00.py::TestSuite::test_timepoint_strftime_strptime PASSED                                                                                                                            [ 47%]
isodatetime/tests/test_00.py::TestSuite::test_timepoint_subtract PASSED                                                                                                                                     [ 50%]
isodatetime/tests/test_00.py::TestSuite::test_timepoint_time_zone PASSED                                                                                                                                    [ 52%]
isodatetime/tests/test_00.py::TestSuite::test_timerecurrence PASSED                                                                                                                                         [ 54%]
isodatetime/tests/test_00.py::TestSuite::test_timerecurrence_alt_calendars PASSED                                                                                                                           [ 56%]
isodatetime/tests/test_00.py::TestSuite::test_timerecurrence_parser PASSED                                                                                                                                  [ 58%]
isodatetime/tests/test_00.py::TestSuite::test_util_cache PASSED                                                                                                                                             [ 60%]
isodatetime/tests/test_datetimeoper.py::TestDateTimeOperator::test_diff_time_point_strs PASSED                                                                                                              [ 62%]
isodatetime/tests/test_datetimeoper.py::TestDateTimeOperator::test_format_duration_str_x PASSED                                                                                                             [ 64%]
isodatetime/tests/test_datetimeoper.py::TestDateTimeOperator::test_process_time_point_str_calendar PASSED                                                                                                   [ 66%]
isodatetime/tests/test_datetimeoper.py::TestDateTimeOperator::test_process_time_point_str_format PASSED                                                                                                     [ 68%]
isodatetime/tests/test_datetimeoper.py::TestDateTimeOperator::test_process_time_point_str_now_0 PASSED                                                                                                      [ 70%]
isodatetime/tests/test_datetimeoper.py::TestDateTimeOperator::test_process_time_point_str_ref_0 PASSED                                                                                                      [ 72%]
isodatetime/tests/test_datetimeoper.py::TestDateTimeOperator::test_process_time_point_str_ref_1 PASSED                                                                                                      [ 75%]
isodatetime/tests/test_datetimeoper.py::TestDateTimeOperator::test_process_time_point_str_ref_2 PASSED                                                                                                      [ 77%]
isodatetime/tests/test_datetimeoper.py::TestDateTimeOperator::test_process_time_point_str_x PASSED                                                                                                          [ 79%]
isodatetime/tests/test_main.py::TestMain::test_1_bad PASSED                                                                                                                                                 [ 81%]
isodatetime/tests/test_main.py::TestMain::test_1_good PASSED                                                                                                                                                [ 83%]
isodatetime/tests/test_main.py::TestMain::test_1_null PASSED                                                                                                                                                [ 85%]
isodatetime/tests/test_main.py::TestMain::test_2_bad PASSED                                                                                                                                                 [ 87%]
isodatetime/tests/test_main.py::TestMain::test_2_good PASSED                                                                                                                                                [ 89%]
isodatetime/tests/test_main.py::TestMain::test_3_bad PASSED                                                                                                                                                 [ 91%]
isodatetime/tests/test_main.py::TestMain::test_3_good PASSED                                                                                                                                                [ 93%]
isodatetime/tests/test_main.py::TestMain::test_4_bad PASSED                                                                                                                                                 [ 95%]
isodatetime/tests/test_main.py::TestMain::test_4_good PASSED                                                                                                                                                [ 97%]
isodatetime/tests/test_time_point.py::TestTimePointCompat::test_timepoint SKIPPED                                                                                                                           [100%]
============================================================================================= short test summary info =============================================================================================
SKIP [1] isodatetime/tests/test_00.py:1055: need --runslow option to run
SKIP [1] isodatetime/tests/test_time_point.py: need --runslow option to run

====================================================================================== 46 passed, 2 skipped in 6.46 seconds =======================================================================================
```

If you would like to activate the slow tests (marked with a pytest decorator), then you can run `python3 setup.py test --pytest-args="--runslow"`. But that will take quite a while... Still need to wait and see a) if Travis-CI will be happy with this change, and b) if Codecov will report the same coverage report as in your change. Going home in a few minutes, but will connect again in 2/3 hours to look at Travis again.

Cheers
Bruno

matthewrmshin commented:
I can get tests to pass in my work environment, although I have not tried running the slow tests in a low power machine yet. Let me know when this is ready for merge. (Sorry I have just pushed a change that conflicts with your branch.)

kinow commented:
Maybe we should merge your changes first, leaving the branch ready for review/merge. And later I will raise other pull requests for pytest+setup.py, maybe functools, etc. WDYT?

I thought adding setup.py and pytest would work out with minimal efgort, but am concerned about blocking your work unnecessarily. WDYT?

matthewrmshin commented:
I don't mind either way. (However, I'd imagine 2nd review will be easier if you raise a separate PR afterwards.)

kinow commented:
Removed `tdqm` dependency and use. Split per change, so now there are 5 commits, as I think it might be easier to remove/edit the commits if necessary :+1: might be easier to review too?

kinow commented:
Passed Travis CI build. Phew :relieved:

kinow commented on setup.py:
```diff
@@ -17,13 +17,31 @@
+
+    def initialize_options(self):
+        TestCommand.initialize_options(self)
+        self.pytest_args = ""
```
Yes. I cannot explain exactly how (spent like 5 minutes searching for it in `setuptools` but couldn't find it), but if you look at the rest of the changes, there's a class variable `user_options`, right below the `class PyTest(TestCommand)`. That defines which properties can be used (other values result in error).

Weirdly, the format for the `--param`, is `--param=`... you can look at the parent class `TestCommand` and you will see other parameters declared like that. `a` is for `-a`.

When you pass a value like `--pytest-args='--runslow'`, the value `--runslow` will be assigned to `self.pytest_args`.

matthewrmshin replied:
Tested + working.

kinow commented on pytest.ini:
```diff
@@ -16,30 +16,8 @@
+addopts = -v -s -ra --color=auto
+# Some tests fail when running in a different time zone (e.g. Pacific/Auckland)
+env =
+    TZ=UTC
```
Done https://github.com/metomi/isodatetime/issues/116

kinow commented on .travis.yml:
```diff
@@ -26,8 +26,13 @@ matrix:
+  # Coverage dependencies, coverage for reporting coverage, pytest-cov to
+  # enable PyTest to produce coverage reports, and pytest-env to allow
+  # PyTest to use environment variables in its configuration (e.g. TZ=UTC).
+  - pip install coverage pytest-cov pytest-env pytest-runner tqdm
```
Done! Edited commit, and push-forced.

kinow commented:
>Some minor comments. Are we happy to merge this now?

All good for me, assuming Travis-CI doesn't fail now (have a tab open on https://travis-ci.org/kinow/isodatetime, might take ~30 mins to finish).
